File: src/main/java/cc/woverflow/hytils/command/HytilsCommand.java
/*
 * Hytils Reborn - Hypixel focused Quality of Life mod.
 * Copyright (C) 2020, 2021, 2022, 2023  Polyfrost, Sk1er LLC and contributors
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package cc.woverflow.hytils.command;

import cc.polyfrost.oneconfig.utils.Multithreading;
import cc.polyfrost.oneconfig.utils.Notifications;
import cc.polyfrost.oneconfig.utils.commands.CommandManager;
import cc.polyfrost.oneconfig.utils.commands.annotations.Command;
import cc.polyfrost.oneconfig.utils.commands.annotations.Description;
import cc.polyfrost.oneconfig.utils.commands.annotations.Main;
import cc.polyfrost.oneconfig.utils.commands.annotations.SubCommand;
import cc.woverflow.hytils.HytilsReborn;
import cc.woverflow.hytils.command.parser.GEXPType;
import cc.woverflow.hytils.command.parser.GEXPTypeParser;
import cc.woverflow.hytils.command.parser.WinstreakType;
import cc.woverflow.hytils.command.parser.WinstreakTypeParser;
import cc.woverflow.hytils.config.HytilsConfig;
import cc.woverflow.hytils.util.HypixelAPIUtils;
import com.mojang.authlib.GameProfile;
import net.minecraft.util.EnumChatFormatting;
import org.jetbrains.annotations.Nullable;

import java.util.Locale;

@Command(value = "hytils", aliases = {"hytilities", "hytilsreborn", "hytilitiesreborn", "hytil"})
public class HytilsCommand {
    static {
        CommandManager.INSTANCE.addParser(new GEXPTypeParser());
        CommandManager.INSTANCE.addParser(new WinstreakTypeParser());
    }

    @Main
    private void handleDefault() {
        HytilsReborn.INSTANCE.getConfig().openGui();
    }

    @SubCommand(description = "Shows your guild experience", aliases = {"guildexp", "guildexperience"})
    @SuppressWarnings("SameParameterValue")
    private void gexp(@Description("username") @Nullable GameProfile player, @Description("type") @Nullable GEXPType type) {
        Multithreading.runAsync(() -> {
            if (HytilsConfig.apiKey.isEmpty() || !HypixelAPIUtils.isValidKey(HytilsConfig.apiKey)) {
                HytilsReborn.INSTANCE.sendMessage(EnumChatFormatting.RED + "You need to provide a valid API key to run this command! Type /api new to autoset a key.");
                return;
            }
            if (player != null) {
                if (type == null) {
                    if (HypixelAPIUtils.getGEXP(player.getName())) {
                        Notifications.INSTANCE
                            .send(HytilsReborn.MOD_NAME, player.getName() + " currently has " + HypixelAPIUtils.gexp + " guild EXP.");
                    } else {
                        Notifications.INSTANCE
                            .send(HytilsReborn.MOD_NAME, "There was a problem trying to get " + player.getName() + "'s GEXP.");
                    }
                } else {
                    switch (type) {
                        case DAILY:
                            if (HypixelAPIUtils.getGEXP(player.getName())) {
                                Notifications.INSTANCE
                                    .send(
                                        HytilsReborn.MOD_NAME,
                                        player.getName() + " currently has " + HypixelAPIUtils.gexp + " daily guild EXP."
                                    );
                            } else {
                                Notifications.INSTANCE
                                    .send(HytilsReborn.MOD_NAME, "There was a problem trying to get " + player.getName() + "'s daily GEXP.");
                            }
                        case WEEKLY:
                            if (HypixelAPIUtils.getWeeklyGEXP(player.getName())) {
                                Notifications.INSTANCE
                                    .send(
                                        HytilsReborn.MOD_NAME,
                                        player.getName() + " currently has " + HypixelAPIUtils.gexp + " weekly guild EXP."
                                    );
                            } else {
                                Notifications.INSTANCE
                                    .send(HytilsReborn.MOD_NAME, "There was a problem trying to get " + player.getName() + "'s weekly GEXP.");
                            }
                    }
                }
            } else {
                if (HypixelAPIUtils.getGEXP()) {
                    Notifications.INSTANCE
                        .send(HytilsReborn.MOD_NAME, "You currently have " + HypixelAPIUtils.gexp + " guild EXP.");
                } else {
                    Notifications.INSTANCE
                        .send(HytilsReborn.MOD_NAME, "There was a problem trying to get your GEXP.");
                }
            }
        });
    }

    @SubCommand(description = "Shows your winstreak", aliases = {"winstreak", "ws"})
    @SuppressWarnings("SameParameterValue")
    private void winstreak(@Description("username") @Nullable GameProfile player, @Description("type") @Nullable WinstreakType gamemode) {
        Multithreading.runAsync(() -> {
            if (HytilsConfig.apiKey.isEmpty() || !HypixelAPIUtils.isValidKey(HytilsConfig.apiKey)) {
                HytilsReborn.INSTANCE.sendMessage(EnumChatFormatting.RED + "You need to provide a valid API key to run this command! Type /api new to autoset a key.");
                return;
            }
            if (player != null) {
                if (gamemode != null) {
                    if (HypixelAPIUtils.getWinstreak(player.getName(), gamemode.name())) {
                        Notifications.INSTANCE
                            .send(
                                HytilsReborn.MOD_NAME,
                                player.getName() + " currently has a " + HypixelAPIUtils.winstreak + " winstreak in " + gamemode.name().toLowerCase(Locale.ENGLISH) + "."
                            );
                    } else {
                        Notifications.INSTANCE
                            .send(
                                HytilsReborn.MOD_NAME,
                                "There was a problem trying to get " + player.getName() + "'s winstreak in $gamemode."
                            );
                    }
                } else {
                    if (HypixelAPIUtils.getWinstreak(player.getName())) {
                        Notifications.INSTANCE
                            .send(
                                HytilsReborn.MOD_NAME,
                                player.getName() + " currently has a " + HypixelAPIUtils.winstreak + " winstreak."
                            );
                    } else {
                        Notifications.INSTANCE
                            .send(HytilsReborn.MOD_NAME, "There was a problem trying to get " + player.getName() + "'s winstreak.");
                    }
                }
            } else {
                if (HypixelAPIUtils.getWinstreak()) {
                    Notifications.INSTANCE
                        .send(HytilsReborn.MOD_NAME, "You currently have a " + HypixelAPIUtils.winstreak + " winstreak.");
                } else {
                    Notifications.INSTANCE
                        .send(HytilsReborn.MOD_NAME, "There was a problem trying to get your winstreak.");
                }
            }
        });
    }

    @SubCommand(description = "Sets your API key.", aliases = "setkey")
    private static void key(@Description("API Key") String apiKey) {
        Multithreading.runAsync(() -> {
            if (HypixelAPIUtils.isValidKey(apiKey)) {
                HytilsConfig.apiKey = apiKey;
                HytilsReborn.INSTANCE.getConfig().save();
                HytilsReborn.INSTANCE.sendMessage(EnumChatFormatting.GREEN + "Saved API key successfully!");
            } else {
                HytilsReborn.INSTANCE.sendMessage(EnumChatFormatting.RED + "Invalid API key! Please try again.");
            }
        });
    }
}

File: src/main/kotlin/org/polyfrost/hytils/client/commands/impl/HytilsCommand.kt
package org.polyfrost.hytils.client.commands.impl

import com.mojang.brigadier.Command
import com.mojang.brigadier.arguments.StringArgumentType
import com.mojang.brigadier.builder.LiteralArgumentBuilder
import net.fabricmc.fabric.api.client.command.v2.ClientCommandManager
import net.fabricmc.fabric.api.client.command.v2.FabricClientCommandSource
import org.polyfrost.hytils.HytilsRebornConstants
import org.polyfrost.hytils.client.HytilsRebornConfig
import org.polyfrost.hytils.client.commands.ClientCommand
import org.polyfrost.hytils.client.commands.parser.GEXPTypeArgumentType
import org.polyfrost.hytils.client.commands.parser.GameArgumentType
import org.polyfrost.hytils.client.utils.HypixelAPIUtils
import org.polyfrost.oneconfig.utils.v1.Multithreading
import org.polyfrost.oneconfig.utils.v1.dsl.mc
import org.polyfrost.oneconfig.utils.v1.dsl.openUI
import java.util.Locale

// TODO: notifications (or just use chat messages)
object HytilsCommand : ClientCommand {
    override fun getCommand(): LiteralArgumentBuilder<FabricClientCommandSource> = ClientCommandManager.literal("hytils")
        .executes { _ ->
            HytilsRebornConfig.openUI()
            Command.SINGLE_SUCCESS
        }
        .then(ClientCommandManager.literal("gexp")
            .executes { _ ->
                Multithreading.submit { executeGEXP(null, null) }
                Command.SINGLE_SUCCESS
            }
            .then(ClientCommandManager.argument("username", StringArgumentType.string())
                .executes { context ->
                    val username = context.getArgument("username", String::class.java)
                    Multithreading.submit { executeGEXP(username, null) }
                    Command.SINGLE_SUCCESS
                }
                .then(ClientCommandManager.argument("type", GEXPTypeArgumentType.gexpType())
                    .executes { context ->
                        val username = context.getArgument("username", String::class.java)
                        val type = context.getArgument("type", GEXPTypeArgumentType.GEXPType::class.java)
                        Multithreading.submit { executeGEXP(username, type) }
                        Command.SINGLE_SUCCESS
                    }
                )
            )
        )
        .then(ClientCommandManager.literal("winstreak")
            .then(ClientCommandManager.argument("username", StringArgumentType.string())
                .then(ClientCommandManager.argument("gamemode", GameArgumentType.gameType())
                    .executes { context ->
                        val username = context.getArgument("username", String::class.java)
                        val game = context.getArgument("gamemode", GameArgumentType.Game::class.java)
                        Multithreading.submit { executeWinstreak(username, game) }
                        Command.SINGLE_SUCCESS
                    }
                )
            )
        )

    override fun getAliases() = listOf("hytilsreborn", "hytil", "hytilities", "hytilitiesreborn")

    private fun executeGEXP(username: String?, type: GEXPTypeArgumentType.GEXPType?) {
        val gexp = when {
            username == null -> HypixelAPIUtils.getGEXP()
            type == GEXPTypeArgumentType.GEXPType.WEEKLY -> HypixelAPIUtils.getWeeklyGEXP(username)
            else -> HypixelAPIUtils.getGEXP(username)
        }

        val label = when (type) {
            GEXPTypeArgumentType.GEXPType.DAILY -> " daily"
            GEXPTypeArgumentType.GEXPType.WEEKLY -> " weekly"
            null -> ""
        }

        if (gexp != null) {
            val subject = if (username != null) "$username currently has" else "You currently have"
//            mc.execute {
//                Notifications.enqueue(
//                    Notifications.Type.Info,
//                    HytilsRebornConstants.NAME,
//                    "$subject $gexp$label guild EXP."
//                )
//            }
        } else {
            val target = if (username != null) "$username's" else "your"
//            mc.execute {
//                Notifications.enqueue(
//                    Notifications.Type.Error,
//                    HytilsRebornConstants.NAME,
//                    "There was a problem trying to get $target$label guild EXP."
//                )
//            }
        }
    }

    private fun executeWinstreak(username: String, game: GameArgumentType.Game) {
        val isSelf = username == mc.user.name

        val winstreak = HypixelAPIUtils.getWinstreak(username, game.name)

        val label = if (!isSelf) {
            " in ${game.name.lowercase(Locale.ROOT)}"
        } else ""

        if (winstreak != null) {
            val subject = if (!isSelf) "$username currently has" else "You currently have"
//            mc.execute {
//                Notifications.enqueue(
//                    Notifications.Type.Info,
//                    HytilsRebornConstants.NAME,
//                    "$subject a $winstreak winstreak$label."
//                )
//            }
        } else {
            val target = if (!isSelf) "$username's" else "your"
//            mc.execute {
//                Notifications.enqueue(
//                    Notifications.Type.Error,
//                    HytilsRebornConstants.NAME,
//                    "There was a problem trying to get $target winstreak."
//                )
//            }
        }
    }
}
